007-design-docs.md

```diff
@@ -0,0 +1,97 @@
+# Design Documentation Proposal
+
+This document is intended to progress discussion about how to write and maintain [design documentation](https://en.wikipedia.org/wiki/Software_design_description) for Strimzi.
+
+The main goal of design documents for Strimzi is to make contributors more effective by forcing all involved to think through the design and gather feedback from others. People often think the point of a design doc is to to teach others about some system or serve as documentation later on. While those can be beneficial side effects, they are not the goal in and of themselves.
+By having a set of design documentation, behaviours of Strimzi as a whole and each of the components are well defined and are agreed upon.
+
+The benefits of adding design documentation to Strimzi are:
+- Design docs/additions could help when there is conflict about the direction of a proposal/code change.
+- It's hard for new contributors to get into the project and understand behaviours of certain components, this may be fixed by extra technical documentation.
+- If just a couple of named contributors moved on to other things, the velocity of the project would drop considerably. Design doc would help in this case too, capturing conventions and behaviour enforcements that normally only get picked up and tweaked at review time.
+
+
+
+## Where is Strimzi today
+
+Currently there are no design documents for Strimzi, the closest thing Strimzi has would probably be the generated docs for classes - these are more user oriented than design oriented. Some classes also have extensive documentation comments found at the top of their files explaining their usage and design.
+
+
+## Proposal
+
+### When to write a design document
+
+It would be appropriate to add a design document when a significant new feature is being added, examples of this being:
+- A new custom resource type being added, a design document explaining how its configuration would be implemented.
+- A new feature to a pre-existing custom resource type, e.g. new component
+- A reworking of an old feature, deprecating the old feature, including if applicable of how a user would migrate.
+
+### Where to write a design document
+
+A design document should be a markdown file located in the `design` directory of the `strimzi-kafka-operator` project.
+Design documents for other Strimzi projects such as Oauth should also be located here as the intent is that the design documents are all colocated and that all components are a part of 'the operator' project.
```

Comment:
I do not think this makes sense - we have several projects with their own repos. We should either have the design folder in each of them with their own designs or we should have a central repo for all designs.

Reply:
I think that is a fair comment. I will amend this.